modules/hub/src/main/Duct.scala

```diff
@@ -16,14 +16,11 @@ trait Duct {
   // implement async behaviour here
   protected val process: Duct.ReceiveAsync
 
-  def !(msg: Any): Unit = atomic { implicit txn =>
-    stateRef.transform {
-      _.map(_ enqueue msg) orElse {
-        Txn.afterCommit { _ => run(msg) }
-        Duct.emptyQueue
-      }
-    }
-  }
+  def !(msg: Any): Unit =
+    if (stateRef.single.getAndTransform {
+      case None => Some(Queue.empty)
+      case Some(q) => Some(q enqueue msg)
+    } isEmpty) run(msg)
```

Comment:
@isaacl @ornicar Is this really correct? I think this run can be delayed for an arbitrary amount of time. For `Tnx.afterCommit` we had that

> handlers will be executed in their registration order

Reply:
Ah ... but it will only call another (potentially racing) run after `postRun`, so should be fine. And below it will only call a racing `run` if the queue is not empty, so that will never race against this line. Seems correct to me, now.